book/src/developers/designs/application_passwords.md

```diff
@@ -0,0 +1,215 @@
+
+(NOTE: We can't assume these DNs are private - I did consider making these
+app=<secret key>,dc=example,dc=com, but client applications may disclose this
+basedn in ui elements).
+
+When a user authenticates the binddn of the account is set to
+spn=user,app=name,dc=example,dc=com. This difference in base DN triggers Kanidm
+to re-route the authentication to the application specific password, rather than
+the UNIX one.
+
+## Kanidm
+
+### Application Entries
+
+A new class for applications will be added that has an associated reference to a
+set of groups that are allowed access.
+
+(NOTE: It could be good to make this a group such that members of the group
+inherit "memberof=application X". This would make presentation of application
+password links in the UI easier since we can then do a search on (memberof &&
+class = application). Alternately we don't need this since we can in-memory
+cache applications and just compare their member sets to the user sessions
+memberof)
+
+Applications need to have service-account like properties where tokens can be
+generated for them. These are optional since an anonymous bind to kanidm and
+searching under the basedn will continue to work.
+
+Application should have a URL reference to help admins identify where the
+application may be located or accessed.
+
+(NOTE: Future, it could be good to allow customisable instructions for users on
+where to go to use their app password?)
+
+(NOTE: To make it easier to manage tokens, it could be good to have a
+"token-capable" class type to prevent duplication this with service-accounts,
+and allows adding tokens to future types).
+
+### Accounts
+
+The user may wish to have multiple passwords per application. Each password may
+have a label to identify it. For example:
+
+```
+  MAIL
+    iphone: abcd...
+    laptop: bcde...
+  HTTP
+    workstation: cdef...
+```
+
+Person accounts will need a new `Attribute::ApplicationPassword` that stores a
+`ValueSetApplicationPassword`. Each value in the set is a new type to manage these
+secrets and their labeling and the references to the applications.
+
+```
+type ApplicationPasswords = BTreeMap<Uuid, (String, Password)>;
+                                     ^        ^         ^
+                                     |        |         +-> The hashed password
+                                     |        +-> Application password label
+                                     +-> Application password UUID
+
+type ApplicationUuid = Uuid;
+
+struct ValueSetApplicationPassword {
+  map: Map<ApplicationUuid, ApplicationPasswords>
+}
+```
+
+Each value in the set is queried by its UUID. This defines the value as
+`Value::ApplicationPassword(Uuid, ApplicationPassword)`. The
+`ApplicationPassword` type implements `PartialEq` so two application passwords
+are equal if their label is equal and they refer to the same application.
+
+(NOTE: We could consider extending the `CredentialType` to add
```

Comment:
I think we don't want to do this, credential type is already pretty overloaded from past design flaws. I think the proposed design here is clearer and going to be nicer :)